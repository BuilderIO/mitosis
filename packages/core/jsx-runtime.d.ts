export * from './src/jsx-types';
