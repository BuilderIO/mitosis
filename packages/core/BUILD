load("@npm//@bazel/typescript:index.bzl", "ts_project")

package(default_visibility = ["//visibility:public"])

ts_project(
    name = "core",
    srcs = glob(
        [
            "src/**/*.ts",
            "src/**/*.tsx",
        ],
        exclude = [
            "**/__tests__/**",
            "**/*.test.ts",
        ],
    ),
    declaration = True,
    preserve_jsx = True,
    # supports_workers = True, # TODO(misko): unclear why enabling this causes tsc to write to source folder
    tsc = "@npm_core//typescript/bin:tsc", # TODO(misko): unclear why this is needed here, but not in other places.
    tsconfig = "//:tsconfig.bazel.json",
    deps = [
        # "@npm_core//:node_modules",
        "@npm_core//acorn",
        "@npm_core//dedent",
        "@npm_core//prettier",
        "@npm_core//lodash",
        "@npm_core//@types/node",
        "@npm_core//@types/babel__core",
        "@npm_core//@types/json5",
        "@npm_core//@types/dedent",
        "@npm_core//@types/prettier",
        "@npm_core//@types/lodash",
        "@npm_core//@types/object-hash",
        "@npm_core//@types/traverse",
        "@npm_core//@types/rollup__plugin-virtual",
        "@npm_core//@builder.io/sdk",
        "@npm_core//@babel/core",
        "@npm_core//@babel/generator",
        "@npm_core//@rollup/plugin-virtual",
        "@npm_core//@angular/compiler",
        "@npm_core//json5",
        "@npm_core//traverse",
        "@npm_core//rollup",
        "@npm_core//csstype",
        "@npm_core//object-hash",
        "@npm_core//axios",
        "@npm_core//liquidjs",
        "@npm_core//vue-template-compiler",
    ],
)
