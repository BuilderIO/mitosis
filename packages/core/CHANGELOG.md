# Change Log

## 0.0.146

### Patch Changes

- 35becd6: Fix: remove all empty text nodes in JSX parser
- f64d9b0: fix: Vue composition API watch deps
- 35becd6: fix: remove redundant {' '} in React generator

## 0.0.145

### Patch Changes

- 9720fb6: Fix: Svelte reactivity in onUpdate dependencies

## 0.0.144

### Patch Changes

- e4f0019: [fix] `preventDefault` in Qwik

## 0.0.142

### Patch Changes

- a4f8311: moved Vue \_classStringToObject logic behind an option

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
