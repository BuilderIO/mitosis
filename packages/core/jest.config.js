module.exports = {
  clearMocks: true,
  coverageDirectory: 'coverage',
  coverageReporters: ['text', 'clover'],
  coverageThreshold: {
    global: {
      branches: 80,
      functions: 80,
      lines: 80,
      statements: 80,
    },
  },
  globals: {
    'ts-jest': {
      extends: './babel.config.js',
      // module: 'commonjs',
    },
  },
  moduleFileExtensions: ['ts', 'tsx', 'js'],
  modulePathIgnorePatterns: ['dist'],
  moduleNameMapper: {
    '@builder.io/mitosis': '<rootDir>',
  },
  notify: true,
  notifyMode: 'always',
  roots: ['<rootDir>'],
  testMatch: ['**/__tests__/*.test.+(ts|tsx)', '**/*.test.+(ts|tsx)'],
  transform: {
    '^.+\\.raw\\.(ts|tsx|js|jsx)$': 'jest-raw-loader',
    '^.+\\.lite\\.(ts|tsx|js|jsx)$': 'jest-raw-loader',
    '^.+\\.(ts|tsx)$': 'ts-jest',
  },
};
