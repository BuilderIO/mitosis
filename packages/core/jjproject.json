{
  "name": "@builder.io/mitosis",
  "targets": {
    "test": {
      "executor": "@nrwl/vite:vitest",
      "options": {
        "project": "@builder.io/mitosis"
      }
    }
  }
}
