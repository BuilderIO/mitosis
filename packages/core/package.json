{
  "name": "@jsx-lite/core",
  "description": "Write components once, run everywhere. Compiles to Vue, React, Solid, and Liquid. Import code from Figma and Builder.io",
  "keywords": [
    "jsx",
    "compiler",
    "no-code",
    "builder.io",
    "figma",
    "react",
    "vue",
    "solid",
    "liquid",
    "shopify",
    "builder",
    "wysiwyg",
    "drag-drop",
    "jsx-lite",
    "angular"
  ],
  "author": {
    "name": "Builder.io",
    "url": "https://www.builder.io"
  },
  "version": "0.0.3",
  "homepage": "https://github.com/BuilderIO/jsx-lite",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/BuilderIO/jsx-lite"
  },
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "release:dev": "npm run build && npm version prerelease --no-git-tag-version && ALLOW_PUBLISH=true npm publish --tag dev",
    "release:patch": "npm run build && npm version patch --no-git-tag-version && ALLOW_PUBLISH=true npm publish",
    "start": "npm run build -- --watch",
    "ts-node": "ts-node -O '{\"module\": \"commonjs\"}'"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "@babel/core": "^7.12.3",
    "@babel/plugin-syntax-jsx": "^7.12.1",
    "@babel/preset-typescript": "^7.12.1",
    "axios": "^0.21.0",
    "axios-cache-adapter": "^2.5.0",
    "chalk": "^4.1.0",
    "csstype": "^3.0.4",
    "dedent": "^0.7.0",
    "json5": "^2.1.3",
    "liquidjs": "^9.6.0",
    "lodash": "^4.17.20",
    "object-hash": "^2.0.3",
    "prettier": "^2.1.2",
    "traverse": "^0.6.6",
    "ts-node": "^9.0.0",
    "typescript": "^4.0.5",
    "vue-template-compiler": "^2.6.12"
  },
  "devDependencies": {
    "@builder.io/sdk": "^1.1.20",
    "@jsx-lite/eslint-plugin": "../eslint-plugin",
    "@testing-library/jest-dom": "^5.11.5",
    "@types/babel__core": "^7.1.12",
    "@types/dedent": "^0.7.0",
    "@types/jest": "^26.0.15",
    "@types/json5": "0.0.30",
    "@types/lodash": "^4.14.165",
    "@types/object-hash": "^1.3.4",
    "@types/prettier": "^2.1.5",
    "@types/traverse": "^0.6.32",
    "jest": "^26.6.3",
    "jest-raw-loader": "^1.0.1",
    "ts-jest": "^26.4.4"
  }
}
