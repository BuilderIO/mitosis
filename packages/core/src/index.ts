export * from './flow';

export type Context<T> = {};

// These compile away
export const useState = <T>(obj: T) => obj;
export const useRef = () => null as any;
export const useContext = <T = { [key: string]: any }>(key: Context<T>): T =>
  null as unknown as T;
export const createContext = <T = { [key: string]: any }>(
  value: T,
): Context<T> => null as unknown as Context<T>;
export const setContext = <T = { [key: string]: any }>(
  key: Context<T>,
  value: Partial<T>,
): void => {};
export const onMount = (fn: () => any) => null as any;
export const onUpdate = (fn: () => any, deps?: any[]) => null as any;
export const onCreate = (fn: () => any) => null as any;
export const onInit = (fn: () => any) => null as any;
export const onUnMount = (fn: () => any) => null as any;
export const useDynamicTag = (fn: () => any) => null as any;
export const onError = (fn: () => any) => null as any;
export const useMetadata = (obj: object) => null;

export * from './parsers/jsx';
export * from './parsers/builder';
export * from './parsers/angular';
export * from './parsers/liquid';
export * from './parsers/context';
export * from './generators/vue';
export * from './generators/angular';
export * from './generators/context/react';
export * from './generators/context/vue';
export * from './generators/react';
export * from './generators/solid';
export * from './generators/liquid';
export * from './generators/builder';
export * from './generators/qwik/index';
export * from './symbols/symbol-processor';
export * from './generators/html';
export * from './generators/svelte';
export * from './generators/stencil';
export * from './generators/mitosis';
export * from './generators/template';
export * from './generators/swift-ui';
export * from './generators/react-native';
export * from './helpers/parse-reactive-script';
export * from './helpers/is-mitosis-node';

export * from './types/mitosis-node';
export * from './types/mitosis-component';
export * from './types/config';

export * from './plugins/compile-away-builder-components';
export * from './plugins/compile-away-components';
export * from './plugins/map-styles';

export * from './targets';
