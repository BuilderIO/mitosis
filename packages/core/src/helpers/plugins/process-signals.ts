import { Node, types } from '@babel/core';
import generate from '@babel/generator';
import { Target } from '../../types/config';
import { Plugin } from '../../types/plugins';
import { babelTransformExpression } from '../babel-transform';
import { capitalize } from '../capitalize';
import { checkIsDefined } from '../nullable';
import { replaceNodes } from '../replace-identifiers';
import { getSignalMitosisImportForTarget, mapSignalType } from '../signals';
import { createCodeProcessorPlugin } from './process-code';

export const replaceSignalSetters = ({
  code,
  nodeMaps,
}: {
  code: string;
  nodeMaps: {
    from: types.Node;
    setTo: types.Expression;
  }[];
}) => {
  for (const { from, setTo } of nodeMaps) {
    code = babelTransformExpression(code, {
      AssignmentExpression(path) {
        if (path.node.operator !== '=') return;

        const lhs = path.node.left;
        const rhs = path.node.right;

        if (!types.isMemberExpression(lhs)) return;
        if (!(types.isObjectExpression(rhs) || types.isIdentifier(rhs))) return;

        const signalAccess = lhs.object;
        if (!types.isMemberExpression(signalAccess)) return;

        if (generate(signalAccess).code !== generate(from).code) return;

        /**
         * Go from:
         *  a.b.c.value.d = e
         *
         * to:
         *  a.b.setC((PREVIOUS_VALUE) => ({ ...PREVIOUS_VALUE, d: e }))
         */
        const setter = types.cloneNode(setTo);

        // TO-DO: replace all `value` references inside of the set logic with `PREVIOUS_VALUE`.
        const prevValueIdentifier = types.identifier('PREVIOUS_VALUE');
        const setFn = types.arrowFunctionExpression(
          [prevValueIdentifier],
          types.objectExpression([
            types.spreadElement(prevValueIdentifier),
            types.objectProperty(lhs.property, rhs),
          ]),
        );
        const setterExpression = types.callExpression(setter, [setFn]);

        path.replaceWith(setterExpression);
      },
    });
  }
  return code;
};

type SignalMapper = {
  getter: (name: string) => types.Expression;
  setter?: (name: string) => types.Expression;
};

/**
 * Processes `Signal` type imports, transforming them to the target's equivalent and adding the import to the component.
 */
export const getSignalTypePlugin =
  ({ target }: { target: Target }): Plugin =>
  () => ({
    json: {
      pre: (json) => {
        createCodeProcessorPlugin((codeType, json) => {
          switch (codeType) {
            // Skip these for now because they break for svelte: `<svelte:element>` becomes `<svelte: element>`.
            // Besides, fairly impossible to endup with a Signal generic there like `<MyComponent<Signal<number>> />`.
            case 'dynamic-jsx-elements':
              return (x) => x;
            default:
              return (code) => {
                if (json.signals?.signalTypeImportName) {
                  return mapSignalType({
                    code,
                    signalImportName: json.signals.signalTypeImportName,
                    target,
                  });
                }

                return code;
              };
          }
        })(json);

        if (json.signals?.signalTypeImportName) {
          json.imports = json.imports || [];
          const signalMappedImport = getSignalMitosisImportForTarget(target);
          if (signalMappedImport) {
            json.imports.push(signalMappedImport);
          }
        }
      },
    },
  });

const getSignalMapperForTarget = (target: Target): SignalMapper => {
  switch (target) {
    case 'svelte':
      return {
        getter: (name) => types.identifier('$' + name),
      };
    case 'react':
    case 'solid':
      return {
        getter: (name) => types.identifier(name),
        setter: (name) => types.identifier('set' + capitalize(name)),
      };
    default:
      // default case: strip the `.value` accessor
      return {
        getter: (name) => types.identifier(name),
      };
  }
};

/**
 * Processes `mySignal.value` accessors for props, context, and state.
 */
export const getSignalAccessPlugin =
  ({ target }: { target: Target }): Plugin =>
  () => ({
    json: {
      pre: (x) => {
        return createCodeProcessorPlugin((_codeType, json) => (code) => {
          const mapSignal = getSignalMapperForTarget(target);
          const nodeMaps: { from: Node; to: Node; setTo: types.Expression | undefined }[] = [];

          for (const propName in json.props) {
            if (json.props[propName].propertyType === 'reactive') {
              const getter = types.memberExpression(
                types.identifier('props'),
                mapSignal.getter(propName),
              );
              const setter = mapSignal.setter
                ? types.memberExpression(types.identifier('props'), mapSignal.setter(propName))
                : undefined;

              nodeMaps.push({
                from: types.memberExpression(
                  types.memberExpression(types.identifier('props'), types.identifier(propName)),
                  types.identifier('value'),
                ),
                to: getter,
                setTo: setter,
              });

              nodeMaps.push({
                from: types.optionalMemberExpression(
                  types.memberExpression(types.identifier('props'), types.identifier(propName)),
                  types.identifier('value'),
                  false,
                  true,
                ),
                to: getter,
                setTo: setter,
              });
            }
          }

          for (const propName in json.context.get) {
            if (json.context.get[propName].type === 'reactive') {
              nodeMaps.push({
                from: types.memberExpression(types.identifier(propName), types.identifier('value')),
                to: mapSignal.getter(propName),
                setTo: mapSignal.setter ? mapSignal.setter(propName) : undefined,
              });
            }
          }

          for (const propName in json.state) {
            if (json.state[propName]?.propertyType === 'reactive') {
              const to = types.memberExpression(
                types.identifier('state'),
                mapSignal.getter(propName),
              );
              const setTO = mapSignal.setter ? mapSignal.setter(propName) : undefined;

              nodeMaps.push({
                from: types.memberExpression(
                  types.memberExpression(types.identifier('state'), types.identifier(propName)),
                  types.identifier('value'),
                ),
                to: to,
                setTo: setTO,
              });

              nodeMaps.push({
                from: types.optionalMemberExpression(
                  types.memberExpression(types.identifier('state'), types.identifier(propName)),
                  types.identifier('value'),
                  false,
                  true,
                ),
                to: to,
                setTo: setTO,
              });
            }
          }

          const filteredNodeMaps = nodeMaps.filter(
            (
              x,
            ): x is {
              from: types.Node;
              to: types.Node;
              setTo: types.Expression;
            } => checkIsDefined(x.setTo),
          );
          // we run state-setter replacement first, because otherwise the other one will catch it.
          if (filteredNodeMaps.length) {
            code = replaceSignalSetters({ code, nodeMaps: filteredNodeMaps });
          }

          const isK = x.name === 'Blocks' && code.includes('props.context?.value');
          if (isK) {
            console.log({ before: code });
          }

          if (nodeMaps.length) {
            code = replaceNodes({ code, nodeMaps });
          }

          if (isK) {
            console.log({ after: code });
          }

          return code;
        })(x);
      },
    },
  });
