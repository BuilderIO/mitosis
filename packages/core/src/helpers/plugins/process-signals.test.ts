import { types } from '@babel/core';
import { capitalize } from '../capitalize';
import { replaceSignalSetters } from './process-signals';

describe(replaceSignalSetters.name, () => {
  test('should replace signal setters', () => {
    const code = `
    props.builderContextSignal.value.content = {
      ...builderContextSignal.content,
      ...newContent,
      data: {     
        ...builderContextSignal.content?.data, 
        ...newContent?.data },
        meta: {
          ...builderContextSignal.content?.meta,
          ...newContent?.meta,
          breakpoints:
          newContent?.meta?.breakpoints ||
          builderContextSignal.content?.meta?.breakpoints,
        }
        };
        
    builderContextSignal.value.rootState = newRootState;
    `;
    const propName = 'builderContextSignal';

    const output = replaceSignalSetters({
      code,
      nodeMaps: [
        {
          from: types.memberExpression(
            types.memberExpression(types.identifier('props'), types.identifier(propName)),
            types.identifier('value'),
          ),
          setTo: types.memberExpression(
            types.identifier('props'),
            types.identifier('set' + capitalize(propName)),
          ),
        },
        {
          from: types.memberExpression(types.identifier(propName), types.identifier('value')),
          setTo: types.identifier('set' + capitalize(propName)),
        },
      ],
    });
    expect(output).toMatchSnapshot();
  });
});
