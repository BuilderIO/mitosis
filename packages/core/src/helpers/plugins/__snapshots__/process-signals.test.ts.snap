// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replaceSignalSetters > should replace signal setters 1`] = `
"props.setBuilderContextSignal(PREVIOUS_VALUE => ({ ...PREVIOUS_VALUE,
  content: { ...builderContextSignal.content,
    ...newContent,
    data: { ...builderContextSignal.content?.data,
      ...newContent?.data
    },
    meta: { ...builderContextSignal.content?.meta,
      ...newContent?.meta,
      breakpoints: newContent?.meta?.breakpoints || builderContextSignal.content?.meta?.breakpoints
    }
  }
}));
setBuilderContextSignal(PREVIOUS_VALUE => ({ ...PREVIOUS_VALUE,
  rootState: newRootState
}))"
`;
